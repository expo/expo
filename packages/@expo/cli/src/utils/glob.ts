import G, { Glob } from 'glob';

/** Finds all matching files. */
export function everyMatchAsync(pattern: string, options: G.IOptions) {
  return new Promise<string[]>((resolve, reject) => {
    const g = new Glob(pattern, options);
    let called = false;
    const callback = (er: Error | null, matched: string[]) => {
      if (called) return;
      called = true;
      if (er) reject(er);
      else resolve(matched);
    };
    g.on('error', callback);
    g.on('end', (matches) => callback(null, matches));
  });
}

/** Bails out early after finding the first matching file. */
export function anyMatchAsync(pattern: string, options: G.IOptions) {
  return new Promise<string[]>((resolve, reject) => {
    const g = new Glob(pattern, options);
    let called = false;
    const callback = (er: Error | null, matched: string[]) => {
      if (called) return;
      called = true;
      if (er) reject(er);
      else resolve(matched);
    };
    g.on('error', callback);
    g.on('match', (matched) => {
      // We've disabled using abort as it breaks the entire glob package across all instances.
      // https://github.com/isaacs/node-glob/issues/279 & https://github.com/isaacs/node-glob/issues/342
      // For now, just collect every match.
      // g.abort();
      callback(null, [matched]);
    });
    g.on('end', (matches) => callback(null, matches));
  });
}

/**
 * Wait some time, then escape...
 * Adding this because glob can sometimes freeze and fail to resolve if any other glob uses `.abort()`.
 */
export function wrapGlobWithTimeout(
  query: () => Promise<string[]>,
  duration: number
): Promise<string[] | false> {
  return new Promise(async (resolve, reject) => {
    const timeout = setTimeout(() => {
      resolve(false);
    }, duration);

    process.on('SIGINT', () => clearTimeout(timeout));

    try {
      resolve(await query());
    } catch (error) {
      reject(error);
    } finally {
      clearTimeout(timeout);
    }
  });
}
