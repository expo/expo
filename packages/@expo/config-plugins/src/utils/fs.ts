import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}

/** Remove a single file (not directory). Returns `true` if a file was actually deleted. */
export function removeFile(filePath: string): boolean {
  try {
    fs.unlinkSync(filePath);
    return true;
  } catch (error: any) {
    // Skip if the remove did nothing.
    if (error.code === 'ENOENT') {
      return false;
    }
    throw error;
  }
}
