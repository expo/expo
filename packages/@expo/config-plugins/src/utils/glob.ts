/**
 * Sort the glob result alphabetically, to ensure results are identical across different devices (Linux/MacOS).
 * Since `glob@9` the results are determined by the OS and not guaranteed to be sorted.
 *
 * @see https://github.com/isaacs/node-glob/issues/576#issuecomment-1972765500
 */
export function withSortedGlobResult(glob: string[]) {
  return glob.sort((a, b) => a.localeCompare(b));
}
