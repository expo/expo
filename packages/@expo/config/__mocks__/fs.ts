module.exports = require('memfs').fs;

// NOTE(cedric): workaround to also mock `node:fs`
jest.mock('node:fs', () => require('fs'));
