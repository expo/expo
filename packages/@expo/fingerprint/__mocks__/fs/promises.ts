import { fs } from 'memfs';
import path from 'path';

/**
 * Further mock `fs.promises` to ensure the parent directory exists.
 * We used to call `fs.promises.mkdtemp(path.join(os.tmpdir(), ...))`, so that we don't have to worry about os.tmpdir() exists.
 */
const origMkdtemp = fs.promises.mkdtemp;
fs.promises.mkdtemp = (
  prefix: string,
  options?: Parameters<typeof origMkdtemp>[1]
): ReturnType<typeof origMkdtemp> => {
  fs.mkdirSync(path.dirname(prefix), { recursive: true });
  return origMkdtemp(prefix, options);
};

module.exports = fs.promises;
