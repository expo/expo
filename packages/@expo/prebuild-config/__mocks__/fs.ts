const fs = require('memfs').fs;

// Add fs.promises.cp polyfill for memfs since it doesn't support it
fs.promises.cp = async (src: string, dest: string, options?: { recursive?: boolean }) => {
  const path = require('path');

  const srcStat = await fs.promises.stat(src);

  if (srcStat.isDirectory()) {
    await fs.promises.mkdir(dest, { recursive: true });
    const entries = await fs.promises.readdir(src, { withFileTypes: true });

    for (const entry of entries) {
      const srcPath = path.join(src, entry.name);
      const destPath = path.join(dest, entry.name);

      if (entry.isDirectory()) {
        await fs.promises.cp(srcPath, destPath, options);
      } else {
        await fs.promises.copyFile(srcPath, destPath);
      }
    }
  } else {
    await fs.promises.copyFile(src, dest);
  }
};

module.exports = fs;

jest.mock('node:fs', () => require('memfs').fs);
