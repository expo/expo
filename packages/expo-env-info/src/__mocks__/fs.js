import { fs, promises } from 'memfs';

const constants = jest.requireActual('fs').constants;

export { fs, promises, constants };
