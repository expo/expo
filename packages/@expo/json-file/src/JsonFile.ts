import { codeFrameColumns } from '@babel/code-frame';
import fs from 'fs';
import JSON5 from 'json5';
import path from 'path';
import { promisify } from 'util';
import writeFileAtomic from 'write-file-atomic';

import JsonFileError, { EmptyJsonFileError } from './JsonFileError';

const writeFileAtomicAsync: (
  filename: string,
  data: string | Buffer,
  options: writeFileAtomic.Options
) => void = promisify(writeFileAtomic);

export type JSONValue = boolean | number | string | null | JSONArray | JSONObject;
export interface JSONArray extends Array<JSONValue> {}
export interface JSONObject {
  [key: string]: JSONValue | undefined;
}

type Defined<T> = T extends undefined ? never : T;

type Options<TJSONObject extends JSONObject> = {
  badJsonDefault?: TJSONObject;
  jsonParseErrorDefault?: TJSONObject;
  cantReadFileDefault?: TJSONObject;
  ensureDir?: boolean;
  default?: TJSONObject;
  json5?: boolean;
  space?: number;
  addNewLineAtEOF?: boolean;
};

const DEFAULT_OPTIONS = {
  badJsonDefault: undefined,
  jsonParseErrorDefault: undefined,
  cantReadFileDefault: undefined,
  ensureDir: false,
  default: undefined,
  json5: false,
  space: 2,
  addNewLineAtEOF: true,
};

/**
 * The JsonFile class represents the contents of json file.
 *
 * It's polymorphic on "JSONObject", which is a simple type representing
 * and object with string keys and either objects or primitive types as values.
 * @type {[type]}
 */
export default class JsonFile<TJSONObject extends JSONObject> {
  file: string;
  options: Options<TJSONObject>;

  static read = read;
  static readAsync = readAsync;
  static parseJsonString = parseJsonString;
  static writeAsync = writeAsync;
  static getAsync = getAsync;
  static setAsync = setAsync;
  static mergeAsync = mergeAsync;
  static deleteKeyAsync = deleteKeyAsync;
  static deleteKeysAsync = deleteKeysAsync;
  static rewriteAsync = rewriteAsync;

  constructor(file: string, options: Options<TJSONObject> = {}) {
    this.file = file;
    this.options = options;
  }

  read(options?: Options<TJSONObject>): TJSONObject {
    return read(this.file, this._getOptions(options));
  }

  async readAsync(options?: Options<TJSONObject>): Promise<TJSONObject> {
    return readAsync(this.file, this._getOptions(options));
  }

  async writeAsync(object: TJSONObject, options?: Options<TJSONObject>) {
    return writeAsync(this.file, object, this._getOptions(options));
  }

  parseJsonString(json: string, options?: Options<TJSONObject>): TJSONObject {
    return parseJsonString(json, options);
  }

  async getAsync<K extends keyof TJSONObject, TDefault extends TJSONObject[K] | null>(
    key: K,
    defaultValue: TDefault,
    options?: Options<TJSONObject>
  ): Promise<Defined<TJSONObject[K]> | TDefault> {
    return getAsync(this.file, key, defaultValue, this._getOptions(options));
  }

  async setAsync(key: string, value: unknown, options?: Options<TJSONObject>) {
    return setAsync(this.file, key, value, this._getOptions(options));
  }

  async mergeAsync(
    sources: Partial<TJSONObject> | Partial<TJSONObject>[],
    options?: Options<TJSONObject>
  ): Promise<TJSONObject> {
    return mergeAsync<TJSONObject>(this.file, sources, this._getOptions(options));
  }

  async deleteKeyAsync(key: string, options?: Options<TJSONObject>) {
    return deleteKeyAsync(this.file, key, this._getOptions(options));
  }

  async deleteKeysAsync(keys: string[], options?: Options<TJSONObject>) {
    return deleteKeysAsync(this.file, keys, this._getOptions(options));
  }

  async rewriteAsync(options?: Options<TJSONObject>) {
    return rewriteAsync(this.file, this._getOptions(options));
  }

  _getOptions(options?: Options<TJSONObject>): Options<TJSONObject> {
    return {
      ...this.options,
      ...options,
    };
  }
}

function read<TJSONObject extends JSONObject>(
  file: string,
  options?: Options<TJSONObject>
): TJSONObject {
  let json;
  try {
    json = fs.readFileSync(file, 'utf8');
  } catch (error: any) {
    assertEmptyJsonString(json, file);
    const defaultValue = cantReadFileDefault(options);
    if (defaultValue === undefined) {
      throw new JsonFileError(`Can't read JSON file: ${file}`, error, error.code, file);
    } else {
      return defaultValue;
    }
  }
  return parseJsonString(json, options, file);
}

async function readAsync<TJSONObject extends JSONObject>(
  file: string,
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  let json;
  try {
    json = await fs.promises.readFile(file, 'utf8');
  } catch (error: any) {
    assertEmptyJsonString(json, file);
    const defaultValue = cantReadFileDefault(options);
    if (defaultValue === undefined) {
      throw new JsonFileError(`Can't read JSON file: ${file}`, error, error.code);
    } else {
      return defaultValue;
    }
  }
  return parseJsonString(json, options);
}

function parseJsonString<TJSONObject extends JSONObject>(
  json: string,
  options?: Options<TJSONObject>,
  fileName?: string
): TJSONObject {
  assertEmptyJsonString(json, fileName);
  try {
    if (_getOption(options, 'json5')) {
      return JSON5.parse(json);
    } else {
      return JSON.parse(json);
    }
  } catch (e: any) {
    const defaultValue = jsonParseErrorDefault(options);
    if (defaultValue === undefined) {
      const location = locationFromSyntaxError(e, json);
      if (location) {
        const codeFrame = codeFrameColumns(json, { start: location });
        e.codeFrame = codeFrame;
        e.message += `\n${codeFrame}`;
      }
      throw new JsonFileError(`Error parsing JSON: ${json}`, e, 'EJSONPARSE', fileName);
    } else {
      return defaultValue;
    }
  }
}

async function getAsync<TJSONObject extends JSONObject, K extends keyof TJSONObject, DefaultValue>(
  file: string,
  key: K,
  defaultValue: DefaultValue,
  options?: Options<TJSONObject>
): Promise<any> {
  const object = await readAsync(file, options);
  if (key in object) {
    return object[key];
  }
  if (defaultValue === undefined) {
    throw new JsonFileError(`No value at key path "${String(key)}" in JSON object from: ${file}`);
  }
  return defaultValue;
}

async function writeAsync<TJSONObject extends JSONObject>(
  file: string,
  object: TJSONObject,
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  if (options?.ensureDir) {
    await fs.promises.mkdir(path.dirname(file), { recursive: true });
  }
  const space = _getOption(options, 'space');
  const json5 = _getOption(options, 'json5');
  const addNewLineAtEOF = _getOption(options, 'addNewLineAtEOF');
  let json;
  try {
    if (json5) {
      json = JSON5.stringify(object, null, space);
    } else {
      json = JSON.stringify(object, null, space);
    }
  } catch (e: any) {
    throw new JsonFileError(`Couldn't JSON.stringify object for file: ${file}`, e);
  }
  const data = addNewLineAtEOF ? `${json}\n` : json;
  await writeFileAtomicAsync(file, data, {});
  return object;
}

async function setAsync<TJSONObject extends JSONObject>(
  file: string,
  key: string,
  value: unknown,
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  // TODO: Consider implementing some kind of locking mechanism, but
  // it's not critical for our use case, so we'll leave it out for now
  const object = await readAsync(file, options);
  return writeAsync(file, { ...object, [key]: value }, options);
}

async function mergeAsync<TJSONObject extends JSONObject>(
  file: string,
  sources: Partial<TJSONObject> | Partial<TJSONObject>[],
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  const object = await readAsync(file, options);
  if (Array.isArray(sources)) {
    Object.assign(object, ...sources);
  } else {
    Object.assign(object, sources);
  }
  return writeAsync(file, object, options);
}

async function deleteKeyAsync<TJSONObject extends JSONObject>(
  file: string,
  key: string,
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  return deleteKeysAsync(file, [key], options);
}

async function deleteKeysAsync<TJSONObject extends JSONObject>(
  file: string,
  keys: string[],
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  const object = await readAsync(file, options);
  let didDelete = false;

  for (let i = 0; i < keys.length; i++) {
    const key = keys[i];
    if (object.hasOwnProperty(key)) {
      delete object[key];
      didDelete = true;
    }
  }

  if (didDelete) {
    return writeAsync(file, object, options);
  }
  return object;
}

async function rewriteAsync<TJSONObject extends JSONObject>(
  file: string,
  options?: Options<TJSONObject>
): Promise<TJSONObject> {
  const object = await readAsync(file, options);
  return writeAsync(file, object, options);
}

function jsonParseErrorDefault<TJSONObject extends JSONObject>(
  options: Options<TJSONObject> = {}
): TJSONObject | void {
  if (options.jsonParseErrorDefault === undefined) {
    return options.default;
  } else {
    return options.jsonParseErrorDefault;
  }
}

function cantReadFileDefault<TJSONObject extends JSONObject>(
  options: Options<TJSONObject> = {}
): TJSONObject | void {
  if (options.cantReadFileDefault === undefined) {
    return options.default;
  } else {
    return options.cantReadFileDefault;
  }
}

function _getOption<TJSONObject extends JSONObject, K extends keyof Options<TJSONObject>>(
  options: Options<TJSONObject> | undefined,
  field: K
): Options<TJSONObject>[K] {
  if (options) {
    if (options[field] !== undefined) {
      return options[field];
    }
  }
  return DEFAULT_OPTIONS[field];
}

function locationFromSyntaxError(error: any, sourceString: string) {
  // JSON5 SyntaxError has lineNumber and columnNumber.
  if ('lineNumber' in error && 'columnNumber' in error) {
    return { line: error.lineNumber, column: error.columnNumber };
  }
  // JSON SyntaxError only includes the index in the message.
  const match = /at position (\d+)/.exec(error.message);
  if (match) {
    const index = parseInt(match[1], 10);
    const lines = sourceString.slice(0, index + 1).split('\n');
    return { line: lines.length, column: lines[lines.length - 1].length };
  }

  return null;
}

function assertEmptyJsonString(json?: string, file?: string) {
  if (json?.trim() === '') {
    throw new EmptyJsonFileError(file);
  }
}
