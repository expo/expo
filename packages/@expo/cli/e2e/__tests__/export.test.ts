/* eslint-env jest */
import JsonFile from '@expo/json-file';
import execa from 'execa';
import fs from 'fs-extra';
import klawSync from 'klaw-sync';
import path from 'path';
import { sync as globSync } from 'glob';

import {
  execute,
  projectRoot,
  getLoadedModulesAsync,
  setupTestProjectAsync,
  bin,
  ensurePortFreeAsync,
} from './utils';

const originalForceColor = process.env.FORCE_COLOR;
const originalCI = process.env.CI;

beforeAll(async () => {
  await fs.mkdir(projectRoot, { recursive: true });
  process.env.FORCE_COLOR = '0';
  process.env.CI = '1';
  process.env._EXPO_E2E_USE_PATH_ALIASES = '1';
});

afterAll(() => {
  process.env.FORCE_COLOR = originalForceColor;
  process.env.CI = originalCI;
  delete process.env._EXPO_E2E_USE_PATH_ALIASES;
});

it('loads expected modules by default', async () => {
  const modules = await getLoadedModulesAsync(`require('../../build/src/export').expoExport`);
  expect(modules).toStrictEqual([
    '@expo/cli/build/src/export/index.js',
    '@expo/cli/build/src/log.js',
    '@expo/cli/build/src/utils/args.js',
    '@expo/cli/build/src/utils/errors.js',
  ]);
});

it('runs `npx expo export --help`', async () => {
  const results = await execute('export', '--help');
  expect(results.stdout).toMatchSnapshot();
});

describe('server', () => {
  beforeEach(() => ensurePortFreeAsync(8081));
  it(
    'runs `npx expo export`',
    async () => {
      const projectRoot = await setupTestProjectAsync('basic-export', 'with-assets');
      // `npx expo export`
      await execa('node', [bin, 'export', '--source-maps', '--dump-assetmap'], {
        cwd: projectRoot,
        env: {
          NODE_ENV: 'production',
          TEST_BABEL_PRESET_EXPO_MODULE_ID: require.resolve('babel-preset-expo'),
          EXPO_USE_FAST_RESOLVER: 'false',
        },
      });

      const outputDir = path.join(projectRoot, 'dist');
      // List output files with sizes for snapshotting.
      // This is to make sure that any changes to the output are intentional.
      // Posix path formatting is used to make paths the same across OSes.
      const files = klawSync(outputDir)
        .map((entry) => {
          if (entry.path.includes('node_modules') || !entry.stats.isFile()) {
            return null;
          }
          return path.posix.relative(outputDir, entry.path);
        })
        .filter(Boolean);

      const metadata = await JsonFile.readAsync(path.resolve(outputDir, 'metadata.json'));

      expect(metadata).toEqual({
        bundler: 'metro',
        fileMetadata: {
          android: {
            assets: [
              {
                ext: 'png',
                path: 'assets/fb960eb5e4eb49ec8786c7f6c4a57ce2',
              },
              {
                ext: 'png',
                path: 'assets/9ce7db807e4147e00df372d053c154c2',
              },
              {
                ext: 'ttf',
                path: 'assets/3858f62230ac3c915f300c664312c63f',
              },
            ],
            bundle: expect.stringMatching(/_expo\/static\/js\/android\/AppEntry-.*\.hbc/),
          },
          ios: {
            assets: [
              {
                ext: 'png',
                path: 'assets/fb960eb5e4eb49ec8786c7f6c4a57ce2',
              },
              {
                ext: 'png',
                path: 'assets/9ce7db807e4147e00df372d053c154c2',
              },
              {
                ext: 'ttf',
                path: 'assets/2f334f6c7ca5b2a504bdf8acdee104f3',
              },
            ],
            bundle: expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-.*\.hbc/),
          },
        },
        version: 0,
      });

      const assetmap = await JsonFile.readAsync(path.resolve(outputDir, 'assetmap.json'));
      expect(assetmap).toEqual({
        '2f334f6c7ca5b2a504bdf8acdee104f3': {
          __packager_asset: true,
          fileHashes: ['2f334f6c7ca5b2a504bdf8acdee104f3'],
          fileSystemLocation: expect.stringMatching(/\/.*\/basic-export\/assets/),
          files: [expect.stringMatching(/\/.*\/basic-export\/assets\/font\.ios\.ttf/)],
          hash: '2f334f6c7ca5b2a504bdf8acdee104f3',
          httpServerLocation: '/assets/assets',
          name: 'font',
          scales: [1],
          type: 'ttf',
        },

        '3858f62230ac3c915f300c664312c63f': {
          __packager_asset: true,
          fileHashes: ['3858f62230ac3c915f300c664312c63f'],
          fileSystemLocation: expect.stringMatching(/\/.*\/basic-export\/assets/),
          files: [expect.stringMatching(/\/.*\/basic-export\/assets\/font\.ttf/)],
          hash: '3858f62230ac3c915f300c664312c63f',
          httpServerLocation: '/assets/assets',
          name: 'font',
          scales: [1],
          type: 'ttf',
        },
        d48d481475a80809fcf9253a765193d1: {
          __packager_asset: true,
          fileHashes: ['fb960eb5e4eb49ec8786c7f6c4a57ce2', '9ce7db807e4147e00df372d053c154c2'],
          fileSystemLocation: expect.stringMatching(/\/.*\/basic-export\/assets/),
          files: [
            expect.stringMatching(/\/.*\/basic-export\/assets\/icon\.png/),
            expect.stringMatching(/\/.*\/basic-export\/assets\/icon@2x\.png/),
          ],
          hash: 'd48d481475a80809fcf9253a765193d1',
          height: 1,
          httpServerLocation: '/assets/assets',
          name: 'icon',
          scales: [1, 2],
          type: 'png',
          width: 1,
        },
      });

      // If this changes then everything else probably changed as well.
      expect(files).toEqual([
        expect.stringMatching(/_expo\/static\/js\/android\/AppEntry-[\w\d]+\.hbc/),
        expect.stringMatching(/_expo\/static\/js\/android\/AppEntry-[\w\d]+\.hbc\.map/),
        expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-[\w\d]+\.hbc/),
        expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-[\w\d]+\.hbc\.map/),
        expect.stringMatching(/_expo\/static\/js\/web\/AppEntry-[\w\d]+\.js/),
        expect.stringMatching(/_expo\/static\/js\/web\/AppEntry-[\w\d]+\.js\.map/),
        'assetmap.json',
        'assets/2f334f6c7ca5b2a504bdf8acdee104f3',
        'assets/3858f62230ac3c915f300c664312c63f',
        'assets/9ce7db807e4147e00df372d053c154c2',
        'assets/assets/font.3858f62230ac3c915f300c664312c63f.ttf',
        'assets/assets/icon.8034d8318b239108719ff3f22f31ef15.png',
        'assets/assets/icon.8034d8318b239108719ff3f22f31ef15@2x.png',

        'assets/fb960eb5e4eb49ec8786c7f6c4a57ce2',
        'debug.html',
        'favicon.ico',
        'index.html',
        'metadata.json',
      ]);
    },
    // Could take 45s depending on how fast npm installs
    120 * 1000
  );

  it(
    'runs `npx expo export --no-bytecode`',
    async () => {
      const projectRoot = await setupTestProjectAsync('basic-export', 'with-assets');

      await execa(
        'node',
        [bin, 'export', '--source-maps', '--no-bytecode', '--dump-assetmap', '--platform', 'ios'],
        {
          cwd: projectRoot,
          env: {
            NODE_ENV: 'production',
            TEST_BABEL_PRESET_EXPO_MODULE_ID: require.resolve('babel-preset-expo'),
            EXPO_USE_FAST_RESOLVER: 'false',
          },
        }
      );

      const outputDir = path.join(projectRoot, 'dist');
      // List output files with sizes for snapshotting.
      // This is to make sure that any changes to the output are intentional.
      // Posix path formatting is used to make paths the same across OSes.
      const files = klawSync(outputDir)
        .map((entry) => {
          if (entry.path.includes('node_modules') || !entry.stats.isFile()) {
            return null;
          }
          return path.posix.relative(outputDir, entry.path);
        })
        .filter(Boolean);

      const metadata = await JsonFile.readAsync(path.resolve(outputDir, 'metadata.json'));

      expect(metadata).toEqual({
        bundler: 'metro',
        fileMetadata: {
          ios: {
            assets: [
              {
                ext: 'png',
                path: 'assets/fb960eb5e4eb49ec8786c7f6c4a57ce2',
              },
              {
                ext: 'png',
                path: 'assets/9ce7db807e4147e00df372d053c154c2',
              },
              {
                ext: 'ttf',
                path: 'assets/2f334f6c7ca5b2a504bdf8acdee104f3',
              },
            ],
            bundle: expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-.*\.js/),
          },
        },
        version: 0,
      });

      const assetmap = await JsonFile.readAsync(path.resolve(outputDir, 'assetmap.json'));
      expect(assetmap).toEqual({
        '2f334f6c7ca5b2a504bdf8acdee104f3': {
          __packager_asset: true,
          fileHashes: ['2f334f6c7ca5b2a504bdf8acdee104f3'],
          fileSystemLocation: expect.stringMatching(/\/.*\/basic-export\/assets/),
          files: [expect.stringMatching(/\/.*\/basic-export\/assets\/font\.ios\.ttf/)],
          hash: '2f334f6c7ca5b2a504bdf8acdee104f3',
          httpServerLocation: '/assets/assets',
          name: 'font',
          scales: [1],
          type: 'ttf',
        },
        d48d481475a80809fcf9253a765193d1: {
          __packager_asset: true,
          fileHashes: ['fb960eb5e4eb49ec8786c7f6c4a57ce2', '9ce7db807e4147e00df372d053c154c2'],
          fileSystemLocation: expect.stringMatching(/\/.*\/basic-export\/assets/),
          files: [
            expect.stringMatching(/\/.*\/basic-export\/assets\/icon\.png/),
            expect.stringMatching(/\/.*\/basic-export\/assets\/icon@2x\.png/),
          ],
          hash: 'd48d481475a80809fcf9253a765193d1',
          height: 1,
          httpServerLocation: '/assets/assets',
          name: 'icon',
          scales: [1, 2],
          type: 'png',
          width: 1,
        },
      });

      // If this changes then everything else probably changed as well.
      expect(files).toEqual([
        expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-[\w\d]+\.js/),
        expect.stringMatching(/_expo\/static\/js\/ios\/AppEntry-[\w\d]+\.js\.map/),
        'assetmap.json',
        'assets/2f334f6c7ca5b2a504bdf8acdee104f3',
        'assets/9ce7db807e4147e00df372d053c154c2',
        'assets/fb960eb5e4eb49ec8786c7f6c4a57ce2',
        'debug.html',
        'favicon.ico',
        'metadata.json',
      ]);

      // Check if the bundle is minified
      const bundlePath = globSync('**/*.js', {
        cwd: path.join(outputDir, '_expo'),
        absolute: true,
      })[0];

      const bundle = await fs.readFile(bundlePath, 'utf8');

      expect(bundle).toMatch('__d(');
      // General check
      expect(bundle.split('\n').length).toBeLessThan(550);
    },
    // Could take 45s depending on how fast npm installs
    120 * 1000
  );
});
