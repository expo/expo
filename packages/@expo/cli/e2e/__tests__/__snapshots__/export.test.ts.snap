// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs \`npx expo export --help\` 1`] = `
"
  Info
    Export the static files of the app for hosting it on a web server

  Usage
    $ npx expo export <dir>

  Options
    <dir>                      Directory of the Expo project. Default: Current working directory
    --output-dir <dir>         The directory to export the static files to. Default: dist
    --dev                      Configure static files for developing locally using a non-https server
    --no-minify                Prevent minifying source
    --no-bytecode              Prevent generating Hermes bytecode
    --max-workers <number>     Maximum number of tasks to allow the bundler to spawn
    --dump-assetmap            Emit an asset map for further processing
    -p, --platform <platform>  Options: android, ios, web, all. Default: all
    -s, --source-maps          Emit JavaScript source maps
    -c, --clear                Clear the bundler cache
    -h, --help                 Usage info
"
`;
